### DISCUSSION TITLE
Should we start working on a Slack interface?

### DISCUSSION THREAD

**Discussion by pragunbhutani** [pragunbhutani/dbt-llm-agent#24] [Ideas]
Hi folks,

As you've seen, at the moment there are two ways to interact with the library:
- You can `pip install` it and use it inside a jupyter notebook
- You can run the streamlit interface and interact with projects locally

However, in the long run I think it would be great to allow companies to run a Slack server and use the library to run a slack chatbot so their users can ask questions on a Slack channel and receive answers in a thread on the same channel.

To do this, I envision the following steps:
- We'll need to use a persistent backend. I don't think we need to build a backend into the bot, but we can have that as a setup requirement where users provide a DB connection string to use as backend.
- On this backend, we'll need a SQL db and a vector database. Another option is to use postgres with pgvector. Supabase could also be a good option because it provides vector search capabilities.
- The app will have to offer a server than can be run on the cloud and connected to Slack.

I'm not sure how the Slack connection bit can be done, we can figure it out. I think 2-3 people could coordinate on bringing this to life.

Your thoughts?

**Comment by yryuvraj** [comment]
Hey @pragunbhutani !

Postgres and pgvector sound like a good option to me. It's a little tedious to set up, but reliable in the long run. We can also use qdrant, an open-source vector database.

I liked the overview of the basic schema you described and went ahead to build a flow of how things would work under the hood.

Flow : 
![image](https://github.com/pragunbhutani/dbt-llm-tools/assets/86550457/f575506e-16cf-4a4f-a03d-25791eb337f4)

Basic Architecture :
![image](https://github.com/pragunbhutani/dbt-llm-tools/assets/86550457/033bab75-ac68-4809-8b69-2369069799e3)

I am new to this, so please do suggest any change(s), if any.

**Reply by codeblech** [reply]
what is a slack server?

**Reply by yryuvraj** [reply]
Something from which we can GET and POST requests for out bot, we'll have to connect it somewhere, right ?

**Reply by pragunbhutani** [reply]
Yes, we'll have to look into packaging a Slack bolt that users can run on their own servers. For the rest of it, I like the way you've documented the flow. I had started writing about the logic [here](https://pragunbhutani.substack.com/p/building-an-ai-data-analyst-with-163) and [here](https://pragunbhutani.substack.com/p/building-an-ai-data-analyst-with-05b).

You'll find that the building blocks for these steps are already there in the library, we essentially need to figure out two things:
1. Create a Slack bolt app to replace the streamlit interface
2. Replace the local tiny db and chroma db with a more persistent db.

Let's talk about these in the kick off meeting tomorrow.

**Comment by pragunbhutani** [comment]
FYI, for those looking to learn the basics of building Slack apps with python, you get [get started here](https://api.slack.com/tools/bolt-python)

**Comment by pragunbhutani** [comment]
I haven't looked into this one in detail, but it looks like [this tutorial](https://us.fixstars.com/case/llm-slackbot) can show you the basic flow of how to build an LLM powered slack chatbot. Like I said, I'm not sure if it's completely correct, but it'll give you a general sense of direction.

**Comment by pragunbhutani** [comment]
The Slack interface has been implemented. :)
